main.go

```diff
@@ -108,71 +106,36 @@ func main() {
 		rlimit = rate.NewLimiter(rate.Limit(options.RateLimit), 10)
 	}
 	state := <-stateChannel
-	for {
-		// Update state when a new one is emitted
-		select {
-		case state = <-stateChannel:
-		case <-ctx.Done():
-			return
-		default:
-		}
-		if rlimit != nil {
-			rlimit.Wait(context.Background())
+
+	queries := make(chan string)
+
+	go func() {
+		for {
+			// Update state when a new one is emitted
+			select {
+			case state = <-stateChannel:
+			case <-ctx.Done():
+				return
+			default:
+			}
```

Comment:
We should add `defer close(queries)` before the loop, otherwise range by queries loop below will never stop in case of `ctx.Done()`